atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockStore.java

```diff
@@ -0,0 +1,155 @@
+ *
+ * "taken for SWEEP (uuid1)" <- - - - -> OPEN < - - - -> "taken for BACKUP (uuid2)"
+ *
+ * acquireLock(REASON) attempts to move the lock state from "OPEN" to "taken because REASON", returning a LockEntry
+ * that holds a unique identifier for that occasion of taking the lock.
+ *
+ * releaseLock(LockEntry) attempts to move the state from "taken because REASON" to "OPEN", but only succeeds if the
+ * LockEntry currently stored matches the one passed in, ensuring that nobody released the lock when we weren't looking.
+ *
+ * Upon creating the PERSISTED_LOCKS_TABLE, we attempt to enter this state machine by populating the table.
+ * If we fail to do this, it's because someone else also created the table, and populated it before we did.
+ * This is actually OK - all we care about is that we're in the state machine _somewhere_.
+ */
+public final class LockStore {
+    private static final Logger log = LoggerFactory.getLogger(LockStore.class);
+
+    private static final String ROW_NAME = "DeletionLock";
+    private final KeyValueService keyValueService;
+
+    public static final LockEntry LOCK_OPEN = ImmutableLockEntry.builder()
+            .rowName(ROW_NAME)
+            .lockId("-1")
+            .reason("Available")
+            .build();
+
+    private LockStore(KeyValueService kvs) {
+        this.keyValueService = kvs;
+    }
+
+    public static LockStore create(KeyValueService kvs) {
+        kvs.createTable(AtlasDbConstants.PERSISTED_LOCKS_TABLE, AtlasDbConstants.GENERIC_TABLE_METADATA);
+        LockStore lockStore = new LockStore(kvs);
+        ensurePersistedLocksTableIsPopulated(kvs, lockStore);
+        return lockStore;
+    }
+
+    private static void ensurePersistedLocksTableIsPopulated(KeyValueService kvs, LockStore lockStore) {
+        if (lockStore.allLockEntries().isEmpty()) {
+            populateTable(kvs);
+        }
+    }
+
+    @VisibleForTesting
+    void populateTable() {
+        populateTable(keyValueService);
+    }
+
+    private static void populateTable(KeyValueService kvs) {
+        CheckAndSetRequest request = CheckAndSetRequest.newCell(
+                AtlasDbConstants.PERSISTED_LOCKS_TABLE,
+                LOCK_OPEN.cell(),
+                LOCK_OPEN.value());
+        try {
+            kvs.checkAndSet(request);
+        } catch (CheckAndSetException e) {
+            // This can happen if multiple LockStores are started at once. We don't actually mind.
+            // All we care about is that we're in the state machine of "LOCK_OPEN"/"LOCK_TAKEN".
+            // It still might be interesting, so we'll log it.
+            List<String> values = e.getActualValues().stream()
+                    .map(v -> new String(v, StandardCharsets.UTF_8))
+                    .collect(Collectors.toList());
+            log.warn("Encountered a CheckAndSetException when creating the LockStore. This means that two "
```

Comment:
May want to consider downgrading to info since this is almost guaranteed to happen on startup in any HA cluster.